src/services/refactors/extractType.ts

```diff
@@ -0,0 +1,154 @@
+            return { edits, renameFilename, renameLocation };
+        }
+    });
+
+    interface Info { isJS: boolean; selection: TypeNode; firstStatement: Statement; typeParameters: ReadonlyArray<TypeParameterDeclaration>; }
+
+    function getRangeToExtract(context: RefactorContext): Info | undefined {
+        const { file, startPosition } = context;
+        const isJS = isSourceFileJS(file);
+        const current = getTokenAtPosition(file, startPosition);
+        const range = createTextRangeFromSpan(getRefactorContextSpan(context));
+
+        const selection = findAncestor(current, (node => node.parent && rangeContainsSkipTrivia(range, node, file) && !rangeContainsSkipTrivia(range, node.parent, file)));
+        if (!selection || !isTypeNode(selection)) return undefined;
+
+        const checker = context.program.getTypeChecker();
+        const firstStatement = Debug.assertDefined(isJS ? findAncestor(selection, isStatementButNotBlockAndHasJSDoc) : findAncestor(selection, isStatementButNotBlock));
+        const typeParameters = checkAndCollectionTypeArguments(checker, selection, firstStatement, file);
+        if (!typeParameters) return undefined;
+
+        return { isJS, selection, firstStatement, typeParameters };
+    }
+
+    function isStatementButNotBlock(n: Node): n is Statement {
+        return n && isStatement(n) && !isBlock(n);
+    }
+
+    function isStatementButNotBlockAndHasJSDoc(n: Node): n is (Statement & HasJSDoc) {
+        return isStatementButNotBlock(n) && hasJSDocNodes(n);
+    }
+
+    function rangeContainsSkipTrivia(r1: TextRange, node: Node, file: SourceFile): boolean {
+        return rangeContainsStartEnd(r1, skipTrivia(file.text, node.pos), node.end);
+    }
+
+    function checkAndCollectionTypeArguments(checker: TypeChecker, selection: TypeNode, statement: Statement, file: SourceFile): TypeParameterDeclaration[] | undefined {
+        let hasError = false;
+        const result: TypeParameterDeclaration[] = [];
+        visitor(selection);
+        return hasError ? undefined : result;
+
+        function visitor(node: Node) {
+            if (isTypeReferenceNode(node)) {
+                if (isIdentifier(node.typeName)) {
+                    const symbol = checker.resolveName(node.typeName.text, node.typeName, SymbolFlags.TypeParameter, /* excludeGlobals */ true);
+                    if (symbol) {
+                        const declaration = cast(first(symbol.declarations), isTypeParameterDeclaration);
+                        if (rangeContainsSkipTrivia(statement, declaration, file) && !rangeContainsSkipTrivia(selection, declaration, file)) {
+                            result.push(declaration);
+                        }
+                    }
+                }
+            }
+            else if (isInferTypeNode(node)) {
+                const conditionalTypeNode = findAncestor(node, n => isConditionalTypeNode(n) && rangeContainsSkipTrivia(n.extendsType, node, file));

```

Comment:
if the selection contains only a syntactically-correct infer type, then I think this code should (1) not set `hasError` but also (2) not push any results into `result`. What happens if you get an empty result list?

Test cases 31-33 show that this works, but I don't understand why.

Reply:
That caused: 
1. find the selections by largest and topmost type node
2. Infer type node contains only a `TypeParameterDeclaration`

in the following case: 
```ts
type A = /*a*/Promise/*b*/
```
The selection is the `TypeReference` which have the same range as the` Identifier``Promise`.

In cases 31 and 32 were set the `hasError ` flag. 
And the case33:
```ts
type Item<T> = T extends (infer /*a*/P/*b*/)[] ? P : never
```
The selection is the `TypeParameter`(maybe, or an `Identifier`), that is a declaration rather a `TypeNode`.